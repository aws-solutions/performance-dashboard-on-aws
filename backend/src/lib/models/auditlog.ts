/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
