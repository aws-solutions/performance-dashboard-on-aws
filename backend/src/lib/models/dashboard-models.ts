export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};