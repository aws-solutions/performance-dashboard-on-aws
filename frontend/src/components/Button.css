.usa-button--base {
  background-color: #565c65 !important;
}
.usa-button--base:hover,
.usa-button--base.usa-button--hover {
  background-color: #3d4551 !important;
}
.usa-button--base:active,
.usa-button--base.usa-button--active {
  background-color: #1b1b1b !important;
}
.usa-button--base:disabled {
  background-color: #c9c9c9 !important;
}
