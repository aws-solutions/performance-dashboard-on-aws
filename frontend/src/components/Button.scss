@import "../styles/base";

.usa-button--outline {
    @include u-text("base", "!important");
    background-color: transparent !important;
    box-shadow: inset 0 0 0 2px color("base") !important;
}
.usa-button--outline:visited {
    @include u-text("base", "!important");
}
.usa-button--outline:hover,
.usa-button--outline.usa-button--hover {
    @include u-text("base-dark", "!important");
    background-color: transparent !important;
    box-shadow: inset 0 0 0 2px color("base-dark") !important;
}
.usa-button--outline:active,
.usa-button--outline.usa-button--active {
    @include u-text("base-dark", "!important");
    background-color: transparent !important;
    box-shadow: inset 0 0 0 2px color("base-dark") !important;
}

.usa-button--outline:disabled,
.usa-button--outline.usa-button--disabled {
    @include u-text("base-light", "!important");
    background-color: transparent !important;
    box-shadow: inset 0 0 0 2px color("base-light") !important;
}
