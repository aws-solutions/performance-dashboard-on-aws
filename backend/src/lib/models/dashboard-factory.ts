import { v4 as uuidv4 } from "uuid";
import { Dashboard, DashboardItem } from "./dashboard-models";
import { User } from "./user-models";
import topicareaFactory from "./topicarea-factory";

const DASHBOARD: string = 'Dashboard';

function createNew(name: string, topicAreaId: string, topicAreaName: string, user: User): Dashboard {
  return {
    id: uuidv4(),
    name,
    topicAreaId,
    topicAreaName,
    createdBy: user.userId,
  };
}

/**
 * Converts a Dashboard to a DynamoDB item.
 */
function toItem(dashboard: Dashboard): DashboardItem {
  let item: DashboardItem = {
    pk: itemId(dashboard.id),
    sk: itemId(dashboard.id),
    type: DASHBOARD,
    dashboardName: dashboard.name,
    topicAreaName: dashboard.topicAreaName,
    topicAreaId: topicareaFactory.itemId(dashboard.topicAreaId),
    description: dashboard.description,
    overview: dashboard.overview,
    createdBy: dashboard.createdBy,
  };
  return item;
}

/**
 * Converts a DynamoDB item into a Dashboard object
 */
function fromItem(item: DashboardItem): Dashboard {
  const id = item.pk.substring(10);
  let dashboard: Dashboard = {
    id,
    name: item.dashboardName,
    topicAreaId: item.topicAreaId.substring(10),
    topicAreaName: item.topicAreaName,
    description: item.description,
    overview: item.overview,
    createdBy: item.createdBy,
  }
  return dashboard;
}

function itemId(id: string): string { return `${DASHBOARD}#${id}` }

export default {
  toItem,
  createNew,
  fromItem,
  itemId,
};
