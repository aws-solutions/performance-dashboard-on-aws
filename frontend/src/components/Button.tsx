/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { KeyboardEvent, MouseEvent, ReactNode } from "react";
import "./Button.scss";

type Variant = "base" | "default" | "outline" | "secondary" | "accent" | "unstyled";

interface Props {
    children: ReactNode;
    onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
    onKeyDown?: (event: KeyboardEvent<HTMLButtonElement>) => void;
    variant?: Variant;
    className?: string;
    testid?: string;
    disabled?: boolean;
    type?: "submit" | "reset" | "button";
    ariaLabel?: string;
    ariaCurrent?: boolean | "time" | "false" | "true" | "page" | "step" | "location" | "date";
    ariaExpanded?: boolean;
    ariaControls?: string;
    disabledToolTip?: string;
}

const Button = React.forwardRef<HTMLButtonElement, Props>((props, ref) => {
    let variantClassName = "";
    switch (props.variant) {
        case "base":
            variantClassName = " usa-button--base";
            break;
        case "outline":
            variantClassName = " usa-button--outline";
            break;
        case "secondary":
            variantClassName = " usa-button--secondary";
            break;
        case "accent":
            variantClassName = " usa-button--accent-cool";
            break;
        case "unstyled":
            variantClassName = " usa-button--unstyled";
            break;
        default:
            variantClassName = " usa-button--base";
            break;
    }

    let additionalClasses = "";
    if (props.className) {
        additionalClasses = ` ${props.className}`;
    }

    const button = (
        <button
            data-testid={props.testid}
            aria-label={props.ariaLabel}
            aria-current={props.ariaCurrent}
            aria-expanded={props.ariaExpanded}
            aria-controls={props.ariaControls}
            className={`usa-button${variantClassName}${additionalClasses}`}
            disabled={props.disabled}
            ref={ref}
            type={props.type}
            onClick={(e) => {
                if (props.onClick) {
                    props.onClick(e);
                }
            }}
            onKeyDown={(e) => {
                if (props.onKeyDown) {
                    props.onKeyDown(e);
                }
            }}
        >
            {props.children}
        </button>
    );

    // show disabled tooltips
    if (props.disabled && props.disabledToolTip) {
        return (
            <span data-toggle="tooltip" data-placement="top" title={props.disabledToolTip}>
                {button}
            </span>
        );
    }

    return button;
});

export default Button;
