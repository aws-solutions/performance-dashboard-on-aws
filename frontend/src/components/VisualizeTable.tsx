/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import React, { useState, MouseEvent } from "react";
import { DatasetType } from "../models";
import { useTranslation } from "react-i18next";
import Alert from "./Alert";
import Button from "./Button";
import Link from "./Link";
import Spinner from "./Spinner";
import TextField from "./TextField";
import { faTimes } from "@fortawesome/free-solid-svg-icons";
import TableWidget from "./TableWidget";
import Dropdown from "./Dropdown";
import DatasetParsingService from "../services/DatasetParsingService";
import PrimaryActionBar from "./PrimaryActionBar";
import { useWindowSize } from "../hooks";
import UtilsService from "../services/UtilsService";

interface Props {
    widgetId: string;
    errors: any;
    register: Function;
    json: Array<any>;
    originalJson: Array<any>;
    headers: Array<string>;
    csvJson: Array<any>;
    datasetLoading: boolean;
    datasetType: DatasetType | undefined;
    onCancel: (event: MouseEvent<HTMLButtonElement>) => void;
    backStep: (event: MouseEvent<HTMLButtonElement>) => void;
    advanceStep: Function;
    fileLoading: boolean;
    processingWidget: boolean;
    fullPreviewButton: JSX.Element;
    fullPreview: boolean;
    previewPanelId: string;
    submitButtonLabel: string;
    sortByColumn?: string;
    sortByDesc?: boolean;
    setSortByColumn: Function;
    setSortByDesc: Function;
    title: string;
    summary: string;
    showTitle: boolean;
    summaryBelow: boolean;
    columnsMetadata: Array<any>;
    significantDigitLabels: boolean;
    displayWithPages: boolean;
    configHeader: JSX.Element;
}

function VisualizeTable(props: Props) {
    const { t } = useTranslation();
    const [showAlert, setShowAlert] = useState(true);
    const windowSize = useWindowSize();
    const isMobile = windowSize.width <= 600;

    const handleSortDataChange = (event: React.FormEvent<HTMLInputElement>) => {
        const target = event.target as HTMLInputElement;
        if (target.value !== "") {
            const sortData = target.value.split("###");
            const header = sortData[0];
            const desc = sortData[1] === "desc";
            props.setSortByColumn(header);
            props.setSortByDesc(desc);
        } else {
            props.setSortByColumn(undefined);
            props.setSortByDesc(undefined);
        }
    };

    return (
        <div className="grid-row">
            <div className="tablet:grid-col-6" hidden={props.fullPreview}>
                <PrimaryActionBar>
                    {props.configHeader}
                    {props.errors.title && (
                        <Alert
                            type="error"
                            message={t("VisualizeTableComponent.ResolveErrors")}
                            slim
                        ></Alert>
                    )}

                    <fieldset className="usa-fieldset">
                        <legend className={`usa-hint ${isMobile ? "grid-col-12" : "grid-col-6"}`}>
                            {t("AddTableScreen.Configure")}
                        </legend>

                        <TextField
                            id="title"
                            name="title"
                            label={t("VisualizeTableComponent.TableTitle")}
                            hint={t("VisualizeTableComponent.TableTitleHint")}
                            error={
                                props.errors.title && t("VisualizeTableComponent.TableTitleError")
                            }
                            required
                            register={props.register}
                        />

                        <div className="usa-checkbox">
                            <input
                                className="usa-checkbox__input"
                                id="display-title"
                                type="checkbox"
                                name="showTitle"
                                defaultChecked={true}
                                ref={props.register()}
                            />
                            <label className="usa-checkbox__label" htmlFor="display-title">
                                {t("AddTextScreen.ShowTitle")}
                            </label>
                        </div>

                        <div className="margin-top-3 grid-col-6">
                            <Dropdown
                                id="sortData"
                                name="sortData"
                                label={t("SortData")}
                                options={DatasetParsingService.getDatasetSortOptions(
                                    props.originalJson,
                                    props.headers,
                                    t,
                                )}
                                onChange={handleSortDataChange}
                                defaultValue={UtilsService.getSortData(
                                    props.sortByColumn,
                                    props.sortByDesc,
                                )}
                                register={props.register}
                            />
                        </div>

                        <div>
                            <label className="usa-label text-bold">{t("TableOptionsLabel")}</label>
                            <fieldset className="usa-fieldset">
                                <legend className="usa-hint">{t("TableOptionsDescription")}</legend>
                                <div className="usa-checkbox">
                                    <input
                                        className="usa-checkbox__input"
                                        id="significantDigitLabels"
                                        type="checkbox"
                                        name="significantDigitLabels"
                                        defaultChecked={false}
                                        ref={props.register()}
                                    />
                                    <label
                                        className="usa-checkbox__label"
                                        htmlFor="significantDigitLabels"
                                    >
                                        {t("SignificantDigitLabels")}
                                    </label>
                                </div>
                                <div className="usa-checkbox">
                                    <input
                                        className="usa-checkbox__input"
                                        id="displayWithPages"
                                        type="checkbox"
                                        name="displayWithPages"
                                        defaultChecked={false}
                                        ref={props.register()}
                                    />
                                    <label
                                        className="usa-checkbox__label"
                                        htmlFor="displayWithPages"
                                    >
                                        {t("DisplayWithPages")}
                                    </label>
                                </div>
                            </fieldset>
                        </div>

                        <TextField
                            id="summary"
                            name="summary"
                            label={t("VisualizeTableComponent.TableSummary")}
                            hint={
                                <>
                                    {t("VisualizeTableComponent.TableSummaryHint")}{" "}
                                    <Link target="_blank" to={"/admin/markdown"} external>
                                        {t("AddTextScreen.ViewMarkdownSyntax")}
                                    </Link>
                                </>
                            }
                            register={props.register}
                            multiline
                            rows={5}
                        />
                        <div className="usa-checkbox">
                            <input
                                className="usa-checkbox__input"
                                id="summary-below"
                                type="checkbox"
                                name="summaryBelow"
                                defaultChecked={false}
                                ref={props.register()}
                            />
                            <label className="usa-checkbox__label" htmlFor="summary-below">
                                {t("VisualizeTableComponent.TableShowSummary")}
                            </label>
                        </div>
                        <br />
                        <br />
                        <hr />
                        <Button
                            variant="outline"
                            type="button"
                            onClick={props.backStep}
                            className="margin-top-1"
                        >
                            {t("BackButton")}
                        </Button>
                        <Button
                            type="submit"
                            disabled={
                                !props.json.length || props.fileLoading || props.processingWidget
                            }
                            className="margin-top-1"
                        >
                            {props.submitButtonLabel}
                        </Button>
                        <Button
                            variant="unstyled"
                            className="text-base-dark hover:text-base-darker active:text-base-darkest margin-top-1"
                            type="button"
                            onClick={props.onCancel}
                        >
                            {t("Cancel")}
                        </Button>
                    </fieldset>
                </PrimaryActionBar>
            </div>

            <section
                className={props.fullPreview ? "tablet:grid-col-12" : "tablet:grid-col-6"}
                aria-label={t("ContentPreview")}
            >
                <div
                    hidden={!props.json.length}
                    className={`${!props.fullPreview ? "margin-left-4 sticky-preview" : ""}`}
                >
                    {isMobile ? <br /> : props.fullPreviewButton}
                    <div id={props.previewPanelId}>
                        {props.datasetLoading ? (
                            <Spinner
                                className="text-center margin-top-6"
                                label={t("LoadingSpinnerLabel")}
                            />
                        ) : (
                            <>
                                {showAlert &&
                                props.datasetType === DatasetType.StaticDataset &&
                                props.csvJson.length ? (
                                    <Alert
                                        type="info"
                                        message={
                                            <div className="grid-row margin-left-6">
                                                <div className="grid-col-11">
                                                    {t(
                                                        "VisualizeTableComponent.TableCorrectDisplay",
                                                    )}{" "}
                                                    <Link
                                                        to="/admin/formatting"
                                                        target="_blank"
                                                        external
                                                    >
                                                        {t("LearnHowToFormatCSV")}
                                                    </Link>
                                                </div>
                                                <div className="grid-col-1">
                                                    <div className="margin-1">
                                                        <Button
                                                            variant="unstyled"
                                                            className="margin-0-important text-base-dark hover:text-base-darker active:text-base-darkest"
                                                            onClick={() => setShowAlert(false)}
                                                            type="button"
                                                            ariaLabel={t("GlobalClose")}
                                                        >
                                                            <FontAwesomeIcon
                                                                icon={faTimes}
                                                                size="sm"
                                                            />
                                                        </Button>
                                                    </div>
                                                </div>
                                            </div>
                                        }
                                        slim
                                    />
                                ) : (
                                    ""
                                )}
                                <TableWidget
                                    id={props.widgetId}
                                    title={props.showTitle ? props.title : ""}
                                    summary={props.summary}
                                    summaryBelow={props.summaryBelow}
                                    data={props.json}
                                    columnsMetadata={props.columnsMetadata}
                                    sortByColumn={props.sortByColumn}
                                    sortByDesc={props.sortByDesc}
                                    significantDigitLabels={props.significantDigitLabels}
                                    displayWithPages={props.displayWithPages}
                                />
                            </>
                        )}
                    </div>
                </div>
            </section>
        </div>
    );
}

export default VisualizeTable;
