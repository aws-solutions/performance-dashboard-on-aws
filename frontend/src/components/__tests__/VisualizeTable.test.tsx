/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import VisualizeTable from "../VisualizeTable";
import { MemoryRouter } from "react-router-dom";
import { DatasetType } from "../../models";
import Button from "../Button";

test("renders the VisualizeTable component", async () => {
    const wrapper = render(
        <VisualizeTable
            errors={[]}
            register={() => {}}
            json={[
                {
                    id: "1",
                    name: "Banana",
                    updatedAt: "2021-11-11",
                },
                {
                    id: "2",
                    name: "Chocolate",
                    updatedAt: "2020-11-11",
                },
                {
                    id: "3",
                    name: "Vanilla",
                    updatedAt: "2019-11-11",
                },
            ]}
            csvJson={[]}
            datasetLoading={false}
            datasetType={DatasetType.DynamicDataset}
            onCancel={() => {}}
            backStep={() => {}}
            advanceStep={() => {}}
            fileLoading={false}
            processingWidget={false}
            fullPreview={false}
            fullPreviewButton={<Button>Full preview</Button>}
            submitButtonLabel="Add Table"
            summaryBelow={false}
            setSortByColumn={() => {}}
            setSortByDesc={() => {}}
            columnsMetadata={[]}
            headers={["1", "2", "3"]}
            originalJson={[
                {
                    id: "1",
                    name: "Banana",
                    updatedAt: "2021-11-11",
                },
                {
                    id: "2",
                    name: "Chocolate",
                    updatedAt: "2020-11-11",
                },
                {
                    id: "3",
                    name: "Vanilla",
                    updatedAt: "2019-11-11",
                },
            ]}
            showTitle={true}
            significantDigitLabels={false}
            displayWithPages={true}
            summary={""}
            title={"Test"}
            configHeader={<></>}
        />,
        { wrapper: MemoryRouter },
    );
    expect(wrapper.container).toMatchSnapshot();
});
