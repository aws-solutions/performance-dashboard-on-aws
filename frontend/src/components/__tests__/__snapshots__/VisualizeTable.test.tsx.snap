// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the VisualizeTable component 1`] = `
<div>
  <div
    class="grid-row"
  >
    <div
      class="tablet:grid-col-6"
    >
      <div
        class="border-base-lighter border-1px shadow-3 z-500 radius-md padding-2 bg-white position-sticky "
      >
        <fieldset
          class="usa-fieldset"
        >
          <legend
            class="usa-hint grid-col-6"
          >
            Configure table content.
          </legend>
          <div
            class="usa-form-group"
            role="contentinfo"
          >
            <label
              class="usa-label text-bold"
              for="title"
            >
              Table title
            </label>
            <div
              class="usa-hint"
              id="title-description"
            >
              Give your table a descriptive title.
            </div>
            <input
              aria-describedby="title-description"
              class="usa-input"
              id="title"
              name="title"
              type="text"
              value=""
            />
          </div>
          <div
            class="usa-checkbox"
          >
            <input
              checked=""
              class="usa-checkbox__input"
              id="display-title"
              name="showTitle"
              type="checkbox"
            />
            <label
              class="usa-checkbox__label"
              for="display-title"
            >
              Show title on dashboard
            </label>
          </div>
          <div
            class="margin-top-3 grid-col-6"
          >
            <div
              class="usa-form-group"
              role="contentinfo"
            >
              <label
                class="usa-label text-bold"
                for="sortData"
              >
                Sort data (optional)
              </label>
              <div
                class="usa-hint"
                id="sortData-description"
              />
              <select
                aria-describedby="sortData-description"
                class="usa-select"
                id="sortData"
                name="sortData"
              >
                <option
                  selected=""
                  value=""
                >
                  Select an option
                </option>
                <option
                  value="1###asc"
                >
                  "1" A-Z
                </option>
                <option
                  value="1###desc"
                >
                  "1" Z-A
                </option>
                <option
                  value="2###asc"
                >
                  "2" A-Z
                </option>
                <option
                  value="2###desc"
                >
                  "2" Z-A
                </option>
                <option
                  value="3###asc"
                >
                  "3" A-Z
                </option>
                <option
                  value="3###desc"
                >
                  "3" Z-A
                </option>
              </select>
            </div>
          </div>
          <div>
            <label
              class="usa-label text-bold"
            >
              Table Options (optional)
            </label>
            <fieldset
              class="usa-fieldset"
            >
              <legend
                class="usa-hint"
              >
                Format your table
              </legend>
              <div
                class="usa-checkbox"
              >
                <input
                  class="usa-checkbox__input"
                  id="significantDigitLabels"
                  name="significantDigitLabels"
                  type="checkbox"
                />
                <label
                  class="usa-checkbox__label"
                  for="significantDigitLabels"
                >
                  Significant digit labels (123K, 1.2M)
                </label>
              </div>
              <div
                class="usa-checkbox"
              >
                <input
                  class="usa-checkbox__input"
                  id="displayWithPages"
                  name="displayWithPages"
                  type="checkbox"
                />
                <label
                  class="usa-checkbox__label"
                  for="displayWithPages"
                >
                  Display with pages
                </label>
              </div>
            </fieldset>
          </div>
          <div
            class="usa-form-group"
            role="contentinfo"
          >
            <label
              class="usa-label text-bold"
              for="summary"
            >
              Table summary (optional)
            </label>
            <div
              class="usa-hint"
              id="summary-description"
            >
              Give your table a summary to explain it in more depth. It can also be read by screen readers to describe the table for those with visual impairments. This field supports markdown.
               
              <a
                class="usa-link action-link"
                href="/admin/markdown"
                target="_blank"
              >
                <span>
                  View Markdown Syntax
                </span>
                <svg
                  aria-hidden="false"
                  aria-label="(opens in a new tab)"
                  class="svg-inline--fa fa-external-link-alt fa-w-16 fa-xs margin-left-05"
                  data-icon="external-link-alt"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                    fill="currentColor"
                  />
                </svg>
              </a>
            </div>
            <textarea
              aria-describedby="summary-description"
              class="usa-textarea"
              id="summary"
              name="summary"
              rows="5"
              style="height: auto;"
            />
          </div>
          <div
            class="usa-checkbox"
          >
            <input
              class="usa-checkbox__input"
              id="summary-below"
              name="summaryBelow"
              type="checkbox"
            />
            <label
              class="usa-checkbox__label"
              for="summary-below"
            >
              Show summary below table
            </label>
          </div>
          <br />
          <br />
          <hr />
          <button
            class="usa-button usa-button--outline margin-top-1"
            type="button"
          >
            Back
          </button>
          <button
            class="usa-button usa-button--base margin-top-1"
            type="submit"
          >
            Add Table
          </button>
          <button
            class="usa-button usa-button--unstyled text-base-dark hover:text-base-darker active:text-base-darkest margin-top-1"
            type="button"
          >
            Cancel
          </button>
        </fieldset>
      </div>
    </div>
    <section
      aria-label="Content preview"
      class="tablet:grid-col-6"
    >
      <div
        class="margin-left-4 sticky-preview"
      >
        <button
          class="usa-button usa-button--base"
        >
          Full preview
        </button>
        <div>
          
          <div
            aria-label="Test"
            class="overflow-x-hidden overflow-y-hidden "
            tabindex="-1"
          >
            <h2
              class="margin-bottom-1"
            >
              Test
              <a
                aria-label="Copy link: Test"
                class="shareButton margin-left-1"
                href="http://localhost/#undefineda"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-link fa-w-16 fa-xs "
                  data-icon="link"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
                    fill="currentColor"
                  />
                </svg>
              </a>
              <span
                class="anchor"
                id="undefineda"
              />
            </h2>
            <div
              class="Markdown usa-prose margin-top-0 margin-bottom-3 tableSummaryAbove textOrSummary"
            />
            <div
              class="overflow-x-hidden overflow-y-hidden"
            >
              <table
                class="usa-table usa-table--borderless"
                role="table"
              >
                <thead>
                  <tr
                    role="row"
                  >
                    <th
                      aria-sort="none"
                      colspan="1"
                      role="columnheader"
                      scope="col"
                      style="min-width: 150px;"
                    >
                      <button
                        aria-label="Toggle SortBy: id"
                        class="usa-button usa-button--unstyled text-no-underline"
                        style="cursor: pointer;"
                        title="Toggle SortBy: id"
                        type="button"
                      >
                        <span>
                          id
                        </span>
                        <svg
                          aria-hidden="true"
                          class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                          data-icon="chevron-down"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                            fill="currentColor"
                          />
                        </svg>
                      </button>
                    </th>
                    <th
                      aria-sort="none"
                      colspan="1"
                      role="columnheader"
                      scope="col"
                      style="min-width: 150px;"
                    >
                      <button
                        aria-label="Toggle SortBy: name"
                        class="usa-button usa-button--unstyled text-no-underline"
                        style="cursor: pointer;"
                        title="Toggle SortBy: name"
                        type="button"
                      >
                        <span>
                          name
                        </span>
                        <svg
                          aria-hidden="true"
                          class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                          data-icon="chevron-down"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                            fill="currentColor"
                          />
                        </svg>
                      </button>
                    </th>
                    <th
                      aria-sort="none"
                      colspan="1"
                      role="columnheader"
                      scope="col"
                      style="min-width: 150px;"
                    >
                      <button
                        aria-label="Toggle SortBy: updatedAt"
                        class="usa-button usa-button--unstyled text-no-underline"
                        style="cursor: pointer;"
                        title="Toggle SortBy: updatedAt"
                        type="button"
                      >
                        <span>
                          updatedAt
                        </span>
                        <svg
                          aria-hidden="true"
                          class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                          data-icon="chevron-down"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                            fill="currentColor"
                          />
                        </svg>
                      </button>
                    </th>
                  </tr>
                </thead>
                <tbody
                  role="rowgroup"
                >
                  <tr
                    role="row"
                  >
                    <th
                      role="cell"
                      scope="row"
                    >
                      1
                    </th>
                    <td
                      role="cell"
                    >
                      Banana
                    </td>
                    <td
                      role="cell"
                    >
                      2021-11-11
                    </td>
                  </tr>
                  <tr
                    role="row"
                  >
                    <th
                      role="cell"
                      scope="row"
                    >
                      2
                    </th>
                    <td
                      role="cell"
                    >
                      Chocolate
                    </td>
                    <td
                      role="cell"
                    >
                      2020-11-11
                    </td>
                  </tr>
                  <tr
                    role="row"
                  >
                    <th
                      role="cell"
                      scope="row"
                    >
                      3
                    </th>
                    <td
                      role="cell"
                    >
                      Vanilla
                    </td>
                    <td
                      role="cell"
                    >
                      2019-11-11
                    </td>
                  </tr>
                </tbody>
              </table>
              <div
                class="grid-row text-base-darker text-italic padding-y-05 padding-right-1"
              >
                <div
                  class="grid-col-6 text-left"
                >
                  <div
                    style="display: inline-flex;"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-download fa-w-16 fa-xs margin-right-1"
                      data-icon="download"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M216 0h80c13.3 0 24 10.7 24 24v168h87.7c17.8 0 26.7 21.5 14.1 34.1L269.7 378.3c-7.5 7.5-19.8 7.5-27.3 0L90.1 226.1c-12.6-12.6-3.7-34.1 14.1-34.1H192V24c0-13.3 10.7-24 24-24zm296 376v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h146.7l49 49c20.1 20.1 52.5 20.1 72.6 0l49-49H488c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                  <div
                    style="display: inline-flex;"
                  >
                    <div
                      class="margin-right-05"
                    >
                      <a
                        class="text-base-darker"
                        download="Test"
                        href="data:text/csv;charset=utf-8,﻿\\"id\\",\\"name\\",\\"updatedAt\\"
\\"1\\",\\"Banana\\",\\"2021-11-11\\"
\\"2\\",\\"Chocolate\\",\\"2020-11-11\\"
\\"3\\",\\"Vanilla\\",\\"2019-11-11\\""
                        target="_self"
                      >
                        Download table
                      </a>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</div>
`;
